examples/bnn.py

```diff
@@ -0,0 +1,129 @@
+import matplotlib
+matplotlib.use('Agg')  # noqa: E402
+import matplotlib.pyplot as plt
+
+import argparse
+
+import numpy as onp
+import jax.numpy as np
+import jax.random as random
+
+import numpyro.distributions as dist
+from numpyro.handlers import sample, seed, substitute, trace
+from numpyro.hmc_util import initialize_model
+from numpyro.mcmc import mcmc
+
+
+"""
+We demonstrate how to use NUTS to do inference on a simple (small)
+Bayesian neural network with two hidden layers.
+"""
+
+
+# the non-linearity we use in our neural network
+def nonlin(x):
+    return np.tanh(x)
+
+
+# a two-layer bayesian neural network with computational flow
+# given by D_X => D_H => D_H => D_Y where D_H is the number of
+# hidden units. (note we indicate tensor dimensions in the comments)
+def model(X, Y, D_H):
+    D_X = X.shape[1]
+    D_Y = 1
+
+    # sample first layer (we put unit normal priors on all weights)
+    w1 = sample("w1", dist.Normal(np.zeros((D_X, D_H)), np.ones((D_X, D_H))))  # D_X D_H
+    z1 = nonlin(np.matmul(X, w1))   # N D_H  <= first layer of activations
+
+    # sample second layer
+    w2 = sample("w2", dist.Normal(np.zeros((D_H, D_H)), np.ones((D_H, D_H))))  # D_H D_H
+    z2 = nonlin(np.matmul(z1, w2))  # N D_H  <= second layer of activations
+
+    # sample final layer of weights and neural network output
+    w3 = sample("w3", dist.Normal(np.zeros((D_H, D_Y)), np.ones((D_H, D_Y))))  # D_H D_Y
+    z3 = np.matmul(z2, w3)  # N D_Y  <= output of the neural network
+
+    # we put a prior on the observation noise
+    prec_obs = sample("prec_obs", dist.Gamma(3.0, 1.0))
+    sigma_obs = 1.0 / np.sqrt(prec_obs)
+
+    # observe data
+    sample("Y", dist.Normal(z3, sigma_obs), obs=Y)
+
+
+# helper function for HMC inference
+def run_inference(model, args, rng, X, Y, D_H):
+    init_params, potential_fn, constrain_fn = initialize_model(rng, model, X, Y, D_H)
+    hmc_states = mcmc(args.num_warmup, args.num_samples, init_params,
+                      sampler='hmc', potential_fn=potential_fn, constrain_fn=constrain_fn)
+    return hmc_states
+
+
+# helper function for prediction
+def predict(model, rng, latents, X, D_H):
+    # we need to expand prec_obs so that the (hmc) samples dimension doesn't
+    # conflict with the data dimension in the final model sample statement
+    latents['prec_obs'] = latents['prec_obs'][:, None, None]
```

Comment:
this is kind of gross :cry:

Reply:
It's actually not as bad as some of the other examples, like baseball. :) What you could also do is remove this line, and just use vmap to batch over the latents (and rngs) instead in your main method. This is really useful when you need to do more complex manipulation for getting the batch shapes in order.

```python
predictions = vmap(lambda samples, rng: predict(model, rng, samples, X_test, D_H))\
     (latents, random.split(rng_predict, args.num_samples))[..., 0]
```

We can thank @fehiepsi for this really useful trick!

Reply:
As an example to illustrate the trick, you can take a look at [this notebook](https://nbviewer.jupyter.org/github/pyro-ppl/numpyro/blob/e07e1b5470e2753c5b8073afccb4a8659eed1ee5/notebooks/Pylgt.ipynb), where I used `vmap` on `forecast` function to get `forecast_marginal`.

Reply:
ah yeah good point, `vmap` to the rescue!

Reply:
although arguably the better solution would be for named tensor dimensions: then the subsitute-a-full-tensor version would have worked too : )